package.json

```diff
@@ -4,6 +4,12 @@
     "description": "Simple HTML5 charts using the canvas element.",
     "version": "3.8.2",
     "license": "MIT",
+    "sideEffects": [
+        "./auto/**/*",
+        "./dist/chart.js",
```

Comment:
Why would this have side effects, but not the `.esm.js` file? If the `core.scale.defaults.js` has side-effects then won't the `.esm.js` file as well?

Reply:
@benmccann 

`sideEffects` field is working only with "empty" imports.

If nothing is imported from Chart.js

```
import 'chart.js'
```

then [setting defaults](https://github.com/chartjs/Chart.js/blob/master/src/core/core.scale.defaults.js) is doesn't make sense

Reply:
I don't understand. `core.scale.defaults` is used by `core.scale`, which would be used by the ESM package

Reply:
@benmccann `import 'chart.js'` loads `dist/chart.js` - it is one file, which doesn't have references to any files (core.scale.defaults).

Reply:
Rollup builds index.js to dist/chart.js, src/core/core.scale.defaults.js in sideEffects, so rollup includes it in bundle.

When user does `import 'chart.js'`, webpack (for example) sees that dist/chart.js doesn't has side-effects, so it not be included in the bundle.

When user imports something from 'chart.js', sideEffects field doesn't matter.

Reply:
I believe `src/core/core.scale.defaults.js` gets bundled into `dist/chart js`. So I would think that `dist/chart.js` has side effects because it has `src/core/core.scale.defaults.js` bundled inside it. I think if we say it doesn't have side effects some bundlers may tree shake out setting the defaults

Is there a way that we can set the defaults in a side-effect free manner so that we don't have to include either file regardless?

Reply:
Side effects work at module level, so dist/chart.js is just one module and it can not be shaken (because of side effects).
For the esm (bundle), I think only the chunks can be shaken if not used and have no side effects. 

(I guess in theory sourcemap could be used to determine the original modules to shake, but I doubt it)

There are other strategies used after module level tree shaking is done.

Reply:
I wonder if there's even a reason to do bundling for the ESM build? Maybe we should just point to the entry in `src/` or copy them from `src/` to `dist/` if needed? Then we could just mark the one small file as having a side effect instead of the whole 10,000 line bundle

Reply:
@benmccann We want to use TypeScript, so the build step will stay anyway. And will not need to add extensions in the import paths.

> Then we could just mark the one small file as having a side effect instead of the whole 10,000 line bundle

The size of the file doesn't matter.

Reply:
If you are worried about that side-effects from "core.scale.defaults" will every time in the bundle, I can calm down you: in PR with tree-shaking this problem is solved, you can test it on my test bench. 

I propose to merge PRs with ESM and tree-shaking first, because they contain many changes, after that back to this PR, make changes and test it again.